src/3dDiff.c

```diff
@@ -0,0 +1,345 @@
+            iarg++; continue;
+        }
+
+        /* left image */
+        if ( strncmp(argv[iarg],"-a",2) == 0) {
+            if (++iarg >= argc) ERROR_exit("Need dset after -a");
+            a_fname = strdup(argv[iarg]);
+            iarg++; continue;
+        }
+
+        /* right image */
+        if ( strncmp(argv[iarg],"-b",2) == 0) {
+            if (++iarg >= argc) ERROR_exit("Need dset after -b");
+            b_fname = strdup(argv[iarg]);
+            iarg++; continue;
+        }
+
+        ERROR_message("ILLEGAL option: %s\n", argv[iarg]) ;
+                suggest_best_prog_option(argv[0], argv[iarg]);
+        exit(1);
+    }
+
+    /* Validate args */
+    if ( !a_fname )
+        ERROR_exit("No dset a supplied!");
+    if ( !b_fname )
+        ERROR_exit("No dset b supplied!");
+    /* tolerance is guaranteed to exist, validation performed above */
+
+    /*   Thanks to PT for simplifying options for me */
+    disp_opt_sum = brutalist + tabular + quiet + tps_report;
+    if ( !disp_opt_sum ){
+        report = 1;
+    }
+    else if ( disp_opt_sum > 1 ){
+       ERROR_exit("Must choose ONLY one of these display opts:"
+                  "  -brutalist, -tabular, -q, -tps_report");
+    }
+    // ... and in any other case, one should have exactly 1 option
+    // entered by the user, and be OK
+
+
+    /* Load the images and check for mutual compatibility*/
+    set_obliquity_report(0); /* We'll check that ourselves below */
+    ds1 = THD_open_dataset(a_fname);
+    if ( ds1 == NULL ) ERROR_exit("Can't open dataset %s", a_fname);
+    DSET_load(ds1); CHECK_LOAD_ERROR(ds1);
+    ds2 = THD_open_dataset(b_fname);
+    if ( ds2 == NULL ) ERROR_exit("Can't open dataset %s", b_fname);
+    DSET_load(ds2); CHECK_LOAD_ERROR(ds2);
+    if ( THD_dataset_mismatch( ds1, ds2 ) ) {
+        ERROR_exit("Mismatch between dsets!\n");
+    }
+    if ( DSET_NVALS(ds1) != DSET_NVALS(ds2) ) {
+        ERROR_exit(
+            "Incompatible time points: dset %s contains %d timepoints, dset %s contains %d timepoints",
+            a_fname, DSET_NVALS(ds1), b_fname, DSET_NVALS(ds2)
+        );
+    }
+    nt = DSET_NVALS(ds1);
+    nv = DSET_NVOX(ds1);
+    total_elements = (int64_t) nt * nv;
+    /* Allocate array for storing each volume's number of differing
+     * elements, so we can print a table later */
+    counts = calloc(nt, sizeof( int ));
+
+    for (int i = 0; i < nt; ++i) {
+        counts[i] = THD_count_diffs(ds1, ds2, i, tol);
+        if ( counts[i] ) {
+            ++total_volumes_differing;
+            total_elements_differing += counts[i];
+        }
+    }
+    frac_elements = (double)total_elements_differing / total_elements;
+    frac_volumes = (double)total_volumes_differing / DSET_NVALS(ds1);
+
+    /* Do the reporting */
+    if ( tabular ) {
+        for (int i = 0; i < nt; ++i ) {
+            printf(
+                "%*d:\t%d\n",
+                (int)ceil(log10(nt)), i, counts[i]
+            );
+        }
+    }
```

Comment:
I don't quite understand tabular output.  

```
$ 3dDiff -a epi_r1+orig'[0]' -b epi_r1+orig'[0]' -tabular
0:	0
0 0 211200 0.00000

$ 3dDiff -a epi_r1+orig'[0]' -b epi_r1+orig'[1]' -tabular
0:	144312
1 144312 211200 0.68330

$ 3dDiff -a epi_r1+orig'[0,1]' -b epi_r1+orig'[2,3]' -tabular
0:	144312
1:	160266
1 2 2 1.00000
```
In the single volume case, the initial zero (before the colon) is confusing until I see the multivolume case---it might be a bit hard for the user to get the main 0/1 output correct, because of the multiline issue.  Perhaps having the brutalist/summary line first would make the most sense, and then one has each [i+1]th line of output for each [i]th volume.  If that change isn't appealing, the current functionality should probably be spelled out more in the help, so user's don't utilize the wrong value.

Reply:
I think I'll spell this out in the help. I'm a little concerned about putting that line first, because in practice that will always be the last thing that the user looks at rather than the first for 4D data of any normal size.